### ISSUE TITLE
Koodikatselmointi - Code Review

### ISSUE THREAD

**Issue by RainSalamander** [tommise/AppointmentSystem#2]
Heroku app tested on 15.10. between 3.30pm and 5pm, Files downloaded 6pm the same day.

### The Code

The folder structure is standard for this course and thus simple to navigate. I would question the decision to split `signup` into its own entity instead of having it included in `auth`, but since the rest of the files reflect this difference it seems fine.  

Many-to-many relations are gracefully implemented using a helper table and models in general are easy to comprehend.

Tiny inconsistency between views-files, most have a separate method for GET and POST where appropriate but auth and signup don't. Not really a problem but bringing the two in line couldn't hurt.  
The views file for the signup page doesn't appear to have an `if request.method == "GET"` statement either, causing the form to validate when the page is initially opened. Doesn't pose an issue at present and may even be intended as the field requirements are displayed to the end user as a result.  

Method names are descriptive, variable names are _mostly_ descriptive.  
I see you're using variable name 't' a lot, which I presume comes from the similarly named one used in the material. It would be a lot better to use names that refer to the object being presently handled (even if single-letter), such as 's' for service or 'a' for appointment.  
Similarly the 'i' you use in many for-loops would benefit from being more specific, if only to distinquish code between things like `for i in employees` and `for i in users`.  

As for the templates, HTML looks clean aside from using `<p></p>` instead of the line break `<br>`, and a couple indentation errors that didn't affect readability.  
I would recommend adding a `lang="en"` tag to the opening \<html> element in the layout file to improve accessibility, otherwise all is well.

I was glad to note no cases of copy-paste code, and only one or two places where a repeating method calls could be replaced by a variable.
Comments are descriptive and a great help throughout the project for both skimming and understanding the code.

#### Documentation
The project functions are well documented, files are spaced and formatted for clarity and any instructions present are easy to follow. No problems here, well done!  

---
### Heroku testing

**I generally followed the User instructions file, testing for both intended behaviour and bad/non-standard inputs.**

The app works, navbar is functional and adapts to window resizing as expected. The statistics pages work without error, though one could point out the GDPR-issues in listing users' names who don't have a reservation (in a non-personal use project this would need a privacy policy agreement). The popular service listing is a cool idea, I like it!
While I did encounter some issues, in most cases form validations were effective at preventing errors, eg. no negative price for a service, no duplicate usernames etc.
The redirect from signup to the login page is a great idea, very user-friendly!

#### Minor Issues
These are things I encountered that, while worth pointing out, don't break the site or pose an urgent problem.

- passwords can only be 20 characters long. Shorter password max lengths may present a vulnerability to bruteforce attacks.
- service names are allowed to be unnececcarily long.
- if a service that has reservations is removed, the reservatios remain. Intended?
- employees can assign appointments without user's consent with appointment updating, intended?
- employees can successfully place reservations by using the reservation URL, probably unintended.
- the 'Update a Service' page doesn't let you update just the name or the price. Additionally the form autofill after not passing validation doesn't only refill the one row 'Update' was clicked on, making the visual feedback confusing. Thankfully only the row that 'Update' is clicked on updates even if all other rows have valid fields, so one won't accidentally overwrite all service names.

#### Major Issues
These are things that I hope will get looked at, as they either don't work as expected or break functionality entirely.

- When adding an appointment the time validation was lacking, but did prevent some bad inputs; though adding a date for 2018 was prevented, past dates were still allowed, and if the time format was invalid the error given was unclear.  

- When reserving an appointment, if you use more than one of the employee/service dropdowns, reservations get placed for all selected services, but the data will be broken as 'user' and 'employee' field values are flipped for the "unselected" reservations. This results in the malformed appointment entries being invisible to the user and un-updateable to the related employee, while still being visible in the 'list all' view.  

- When updating an appointment with a reservation, the 'user' and 'employee' field data gets swapped here too, with the same results as above. I'm sorry that I can't offer more help as to the cause of this behaviour, but it likely lies in the respective methods for the misbehaving pages.  

- Additionally for any appointment update, if the time is changed into an invalid value such as 33:00 the update request results in an internal server error. This only happens when updating, creating is unaffected.  

----
**In summary**, the project seems to be doing okay. Fixing the major bugs should be highest priority and improving any vague variable naming might be beneficial for code quality grading.

#### Thoughts for future development  
- Displaying unreserved appointment times and/or listing available services while not logged in could improve user experience.
- Regular users being able to see service popularity statistics could be a neat addition too.
- User info editing would be good in case someone changes names etc.  
- Prices are currently in full euros only, perhaps implementing cents could be considered?  
- Are all employees able to do all services? Having service options tied to individual employees could be a rework to consider.
